pallets/asset-manager/src/lib.rs

```diff
@@ -0,0 +1,393 @@
+    #[pallet::error]
+    pub enum Error<T> {
+        ErrorCreatingAsset,
+        AssetAlreadyExists,
+        AssetDoesNotExist,
+        TooLowNumAssetsWeightHint,
+    }
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(crate) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// New asset with the asset manager is registered
+        AssetRegistered {
+            asset_id: T::AssetId,
+            asset: T::AssetType,
+            metadata: T::AssetRegistrarMetadata,
+        },
+        /// Changed the amount of units we are charging per execution second for a given asset
+        UnitsPerSecondChanged {
+            asset_type: T::AssetType,
+            units_per_second: u128,
+        },
+        /// Changed the xcm type mapping for a given asset id
+        AssetTypeChanged {
+            asset_id: T::AssetId,
+            new_asset_type: T::AssetType,
+        },
+        /// Removed all information related to an assetId
+        AssetRemoved {
+            asset_id: T::AssetId,
+            asset_type: T::AssetType,
+        },
+        /// Supported asset type for fee payment removed
+        SupportedAssetRemoved { asset_type: T::AssetType },
+    }
+
+    /// Mapping from an asset id to asset type.
+    /// This is mostly used when receiving transaction specifying an asset directly,
+    /// like transferring an asset from this chain to another.
+    #[pallet::storage]
+    #[pallet::getter(fn asset_id_type)]
+    pub type AssetIdType<T: Config> = StorageMap<_, Blake2_128Concat, T::AssetId, T::AssetType>;
+
+    /// Reverse mapping of AssetIdType. Mapping from an asset type to an asset id.
+    /// This is mostly used when receiving a multilocation XCM message to retrieve
+    /// the corresponding asset in which tokens should me minted.
+    #[pallet::storage]
+    #[pallet::getter(fn asset_type_id)]
+    pub type AssetTypeId<T: Config> = StorageMap<_, Blake2_128Concat, T::AssetType, T::AssetId>;
+
+    /// Stores the units per second for local execution for a AssetType.
+    /// This is used to know how to charge for XCM execution in a particular
+    /// asset
+    /// Not all assets might contain units per second, hence the different storage
+    #[pallet::storage]
+    #[pallet::getter(fn asset_type_units_per_second)]
+    pub type AssetTypeUnitsPerSecond<T: Config> =
+        StorageMap<_, Blake2_128Concat, T::AssetType, u128>;
+
+    // Supported fee asset payments
+    #[pallet::storage]
+    #[pallet::getter(fn supported_fee_payment_assets)]
+    pub type SupportedFeePaymentAssets<T: Config> = StorageValue<_, Vec<T::AssetType>, ValueQuery>;
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Register new asset with the asset manager
+        #[pallet::weight(T::WeightInfo::register_asset())]
+        pub fn register_asset(
+            origin: OriginFor<T>,
+            asset: T::AssetType,
+            metadata: T::AssetRegistrarMetadata,
```

Comment:
feel asset metadata should be handled by pallet_assets, normally we just need `assetId` and `multiLocation` here

Reply:
assetId is uniquely [generated](https://github.com/rocketman1997/parallel-1/blob/9508db5cf074b53e4ae3d18618d6f2614a479533/runtime/heiko/src/lib.rs#L1001-L1012) from `Asset_Type` and during `register_asset` will [create asset](https://github.com/rocketman1997/parallel-1/blob/9508db5cf074b53e4ae3d18618d6f2614a479533/runtime/heiko/src/lib.rs#L1018-L1043) internally with the provided metadata in one step so to simplifying processes, or need 3-4 calls seperately to register a foreign asset

Reply:
some revamp here to allow register already exist asset with provided asset_id, if so just ignore creating asset